drivers: video: Add ov5642 camera driver

Hardevsinh-Palaniya opened the pull request:
Add driver to support ov5642 camera sensor

avolmat-st commented:
Thanks @hardevsinh-1 for this PR.

There is already an existing driver for the OV5640 sensor. Having very very rapidly checked the difference between the OV5642 and the OV5640, it seems that the OV5642 has less features than the OV5640, having for example only DVP support and limited to 720p maximum frames.
However, considering the numbering of the sensor I am wondering if the sensor behavior might be close enough to the OV5640 so that we could reuse the OV5640 sensor driver to also drive this new OV5642 sensor.
If that is so, and OV5642 is a "subset" of the OV5640, we might be able to make some of the OV5640 features optional or at least checked at runtime if applicable and have both OV5640 / OV5642 handled by the same driver.

Do you know the differences between the OV5640 and the OV5642 and the difference in the way to control it ? Looking very rapidly, it seems that at least on some aspects it is very similar to the OV5640.

Of course if the sensor are very different or if the way to control them is very different, then it makes sense to have a different driver.

hardevsinh-1 commented:
> Thanks @hardevsinh-1 for this PR.
> 
> There is already an existing driver for the OV5640 sensor. Having very very rapidly checked the difference between the OV5642 and the OV5640, it seems that the OV5642 has less features than the OV5640, having for example only DVP support and limited to 720p maximum frames. However, considering the numbering of the sensor I am wondering if the sensor behavior might be close enough to the OV5640 so that we could reuse the OV5640 sensor driver to also drive this new OV5642 sensor. If that is so, and OV5642 is a "subset" of the OV5640, we might be able to make some of the OV5640 features optional or at least checked at runtime if applicable and have both OV5640 / OV5642 handled by the same driver.
> 
> Do you know the differences between the OV5640 and the OV5642 and the difference in the way to control it ? Looking very rapidly, it seems that at least on some aspects it is very similar to the OV5640.
> 
> Of course if the sensor are very different or if the way to control them is very different, then it makes sense to have a different driver.

@avolmat-st, Thanks for the review.
After comparing the OV5640 and OV5642 sensors and their drivers, here are the key points:

- Register maps differ (clock, PLL, exposure, gain, and ISP controls).
- OV5642 lacks AF and advanced ISP blocks present in OV5640.
- DVP timing parameters also differ, and applying OV5640’s settings leads to jitter or frame corruption.
- The initialization flow and power-up timings differ , reusing OV5640’s sequence causes I²C lockups. Sharing one driver would add unnecessary complexity.

Given these differences, what would you recommend ?

josuah reviewed (commented):
Thank you for this new driver!

Image sensor drivers are almost just configuration data.

I agree that it is worth considering de-duplicating all the code with ~~`ov5640.c`~~ `video_common.c`.

It might be possible to reduce the driver by a fair 50% once all video APIs are in place.

It is maybe surprising that copying an existing image sensor driver as much as possible leads to a lot of requests for modifications, but this is because image sensor currently in Zephyr almost all need a round of refactoring and I am very late on this!

Finally, you would need to add an extra commit to add this sensor to the `build_all` test:
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/video/app.overlay

josuah reviewed (commented):
Thank you for the changes!
It seems almost ready, with just a few minor coding style tuning but otherwise seems ready to me.
One more cycle to go!

hardevsinh-1 commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,901 @@
+			       .step = 1,
+			       .def = 0
+			       });
+}
```
Thanks for the note. I will take care when it will be implemented.

josuah reviewed (commented):
Thank you for the edits!
I think there might be one more round needed as I changed my mind after doing other reviews and gave wrong instructions last time (sorry!).

Maybe other reviewers will find other things after that but for my part I think that's all of it after this round.

hardevsinh-1 commented:
> Thank you for the edits! I think there might be one more round needed as I changed my mind after doing other reviews and gave wrong instructions last time (sorry!).
> 
> Maybe other reviewers will find other things after that but for my part I think that's all of it after this round.

Thanks @josuah , I’m interested in adjusting these lines. I’ve made changes based on your suggestions. Could you please review them?

josuah reviewed (commented):
Thank you for the edits! The driver looks ready now.
But there is a "merge commit" that appeared, and you need to remove it.
Then this driver can be merged I think.

josuah commented:
Good catch for `chip_id = tmp`.

I think you still need to remove the merge commits:

![screenshot_2025_1024_1208_42](https://github.com/user-attachments/assets/29ac72d8-d5b9-41a4-b446-e0b8c22a6dc7)

If you wish to catch-up with latest Zephyr `main` branch, you need to avoid `git merge` and instead use `git rebase`, or it will create "merge commits" like seen above.

Now you would need to remove them, for instance with `git reset --hard HEAD~2` and then `git fetch origin && git rebase origin/main` (not tested, just to illustrate).

Feel free to reach on the Discord chat if you wanted to troubleshoot this interactively https://discord.com/invite/Ck7jw53nU2

Hardevsinh-Palaniya committed:
drivers: video: Add ov5642 camera driver

Add driver to support ov5642 camera sensor

Co-developed-by: Rutvij Trivedi <rutvij.trivedi@siliconsignals.io>
Signed-off-by: Rutvij Trivedi <rutvij.trivedi@siliconsignals.io>
Signed-off-by: Hardevsinh Palaniya <hardevsinh.palaniya@siliconsignals.io>

hardevsinh-1 commented:
Hi @josuah ,
Need your input. I got failure in compliance checks at the point which is not changed in my PR. Do I need to update the untouched part of the code to resolve the compliance checks?

josuah reviewed (dismissed):
Thank you for the adjustments, looking all good to me!

josuah commented:
We are now entering [feature-freeze](https://github.com/zephyrproject-rtos/zephyr/wiki/Release-Management), which means this might not be merged in `main` until 4.3.0 is released.

If you need this for your own projects, I would offer you to use `west patch`:

```
west patch gh-fetch -pr 97106 -m zephyr
```

in each of the projects that need it, and then you may use:

```
west patch apply
```

to patch `~/zephyrproject/zephyr`.

Of course it is also possible to use it out-of-tree.

ngphibang commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,908 @@
+enum ov5642_frame_rate {
+	OV5642_15_FPS = 15,
+	OV5642_30_FPS = 30,
+};
```
nit: This could be removed as not used at all. In the future, if the driver supports frame rate setting / enumeration, we could add it and I suggest to place the higher frame rates first. This way applications (e.g. like gstreamer) can choose the highest frame rate as the default one when no particular frame rate is specified.

```
enum ov5642_frame_rate {
	OV5642_15_FPS = 30,
	OV5642_30_FPS = 15,
};
```

ngphibang commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,908 @@
+	const struct ov5642_mode_config *cur_mode;
+};
+
+static const struct video_reg16 init_params_common[] = {
```
nit: It seems you named only the registers which are used later. Naming more registgers may help to have a rough understand what they are even they are used only once in init phase ?

ngphibang commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,908 @@
+		return video_modify_cci_reg(&cfg->i2c,
+				ISP_CTRL01_CCI,
+				BIT(0),
+				ctrls->awb.val ? BIT(0) : 0);
```
I think it should be rather `VIDEO_CID_AUTO_WHITE_BALANCE`, the `VIDEO_CID_WHITE_BALANCE_TEMPERATURE` is to specify the white balance temperature (manual)

ngphibang reviewed (approved):
Thank you for adding a new camera driver. LGTM in overal, just some nits and the VIDEO_CID_AUTO_WHITE_BALANCE comment.

ngphibang reviewed (changes requested):
Apologies for the confusion, I mistakenly approved the PR, but some changes will be needed, especially the AWB CID. You need also fix the compliance check warning:

> author name (Hardevsinh Palaniya) and email (hardevsinh.palaniya@siliconsignals.io) needs to match one of the signed-off-by entries.

ngphibang commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,984 @@
+enum ov5642_frame_rate {
+	OV5642_30_FPS = 30,
+	OV5642_15_FPS = 15,
+};
```
In my previous comment, I meant, in this version, frame rate changing is not supported so this can be dropped, when it is supported, we can place the highest frame rates first ...

Looking at the current code, no set/get/enum_frmival() API is implemented so this can be dropped. Or if this is some thing you can support now (change frame frate between 30 and 15 fps) ? If so, you can implement the above APIs and may take a look at https://github.com/zephyrproject-rtos/zephyr/pull/98154#discussion_r2463040923 as a reference.

ngphibang reviewed (commented):
Thank you for the updates ! I left some comments. The commits seems need to be reorganized, the 2nd and 3rd commits does not reflect their message and titles (they are just fixups) and can be squashed to the 1st one.

hardevsinh-1 commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,984 @@
+
+#define AEC_PK_MANUAL		0x3503
+#define AEC_PK_CCI		OV5642_REG8(AEC_PK_MANUAL)
+#define REAL_GAIN		0x52a6
```
Indentation in my local branch is proper, and while I am trying to view the file in the GitHub editor, the indentation seems proper. Could you please suggest a way to do it properly here?

avolmat-st commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,979 @@
+struct ov5642_data {
+	struct ov5642_ctrls ctrls;
+	struct video_format fmt;
+	uint16_t cur_frmrate;
```
cur_frmrate and cur_mode variable seems never used. Is that is so, should be removed.
Actually cur_mode is set but never used afterward.

avolmat-st commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,979 @@
+	{0x5687, 0xC0},
+};
+
+static const struct ov5642_mode_config dvp_modes[] = {
```
This table is used with the assumption that it has the same number of entries and in the same ordering as the caps. Having some protection around this would be great whenever using it, and at least having a comment before that mentioning that order should not be changed would help preventing futur bugs.

avolmat-st commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,979 @@
+	double rad_val = value;
+	int ret;
+
+	ret = video_modify_cci_reg(&cfg->i2c, SDE_CTRL0_CCI, BIT(0),
```
Nit: seems this could fit in one line. Same in other places below. Line maximum length is 100 caracteres.

ngphibang commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,979 @@
+	int ret;
+
+	ret = video_modify_cci_reg(&cfg->i2c, SDE_CTRL0_CCI, BIT(0),
+				   BIT(0));
```
Yes, there is a a PR but this part is not integrated. That's a good idea, I will add it. The function should be called in `set_stream()` when the HW was fully initialized, and when the HW is was stopped / started, I think.

avolmat-st replied:
> Yes, there is a a PR but this part is not integrated. That's a good idea, I will add it. The function should be called in `set_stream()` when the HW was fully initialized, and when the HW is was stopped / started, I think.

Hum, I was more thinking about calling this at init time to put the sensor in the condition reported by the API. It shouldn't be necessary to call again at set_stream time since ctrls are later on applied when the application calls set_ctrl in fact.

josuah commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,979 @@
+	int ret;
+
+	ret = video_write_cci_reg(&cfg->i2c, SYS_CTRL0_CCI, val);
+	if (ret) {
```
It is related to MISRA (Zephyr 85 here) AFAIR:
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html

avolmat-st reviewed (commented):
Thanks for the update. LGTM.
I raised a small point regarding the k_sleep during the powerup sequence, could be fixed I think but that's optional.
I think one point regarding the white balance control CID remains ?

josuah commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,908 @@
+	case VIDEO_CID_HUE:
+		return ov5642_set_ctrl_hue(dev, ctrls->hue.val);
+	default:
+			return -ENOTSUP;
```
Then maybe this could become `CODE_UNREACHABLE` to help with this?

avolmat-st reviewed (dismissed):
Thanks @hardevsinh-1 for all the fixes.
I was waiting for the CI check to finish.
This look good to me.

ngphibang reviewed (commented):
Thanks for the update ! LGTM nearly except the comment on `VIDEO_CID_EXPOSURE` (which should be `VIDEO_CID_EXPOSURE_AUTO`). Sorry not sure why I was missing this when commenting on the similar VIDEO_CID_AUTO_WHITE_BALANCE ...

Hardevsinh-Palaniya commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,955 @@
+		return video_modify_cci_reg(&cfg->i2c,
+				AEC_PK_CCI,
+				BIT(0),
+				ctrls->ae.val ? BIT(0) : 0);
```
For now, I haven't implemented and tested the manual exposure.

Hardevsinh-Palaniya commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,948 @@
+					    ctrls->vflip.val ? BIT(5) : 0);
+	case VIDEO_CID_EXPOSURE_AUTO:
+		return video_modify_cci_reg(&cfg->i2c,
+					    AEC_PK_CCI, BIT(0), BIT(0));
```
Thanks for the support. I have updated the PR according to the suggestion. Please review it and let me know if anything needs to be updated.

ngphibang commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,969 @@
+					   AEC_PK_CCI, BIT(0), BIT(0));
+		break;
+	default:
+		return -EINVAL;
```
These two cases need to be added as "unsupported" ?

```
case:
	VIDEO_EXPOSURE_SHUTTER_PRIORITY:
	VIDEO_EXPOSURE_APERTURE_PRIORITY:
	         return -ENOTSUP;
```

then the default case:

```
	default:
		CODE_UNREACHABLE;
```

as the video framework already handles values out of the control value range.

ngphibang reviewed (dismissed):
LGTM. Thanks !

avolmat-st commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,970 @@
+		{SDE_CTRL2_REG, abs(sin_coef)}
+	};
+
+	ret = video_modify_cci_reg(&cfg->i2c, SDE_CTRL10_CCI, 0x7F,
```
Nitpicking: can fit in one line.

avolmat-st commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,970 @@
+		return ret;
+	}
+
+	ret = video_modify_cci_reg(&cfg->i2c, SDE_CTRL10_CCI, BIT(2),
```
Nitpicking: can fit in one line.

avolmat-st commented on drivers/video/ov5642.c:
```diff
@@ -0,0 +1,970 @@
+
+	switch (id) {
+	case VIDEO_CID_HFLIP:
+		return video_modify_cci_reg(&cfg->i2c,
```
Nitpicking: can fit in two lines ? same below ?

avolmat-st reviewed (dismissed):
Hi,
sorry for the delay. This looks good to me. Few multi-lines calls could be done in single line but that's really nitpicking so I am all fine with the current code as well.
Thanks.

avolmat-st reviewed (dismissed):
Thanks. Approved.

Hardevsinh-Palaniya commented on drivers/video/CMakeLists.txt:
```diff
@@ -31,6 +31,9 @@ zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_JPEG video_stm32_jpeg.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_ST_MIPID02 video_st_mipid02.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_SW_GENERATOR video_sw_generator.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_JPEG	video_stm32_jpeg.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_HIMAX_HM01B0  hm01b0.c)
```
Thanks @ngphibang , It was a rebase issue. I resolved it.
